master/internal/stream/projects.go

```diff
@@ -0,0 +1,231 @@
+
+// ProjectMsg is a stream.Msg.
+type ProjectMsg struct {
+	bun.BaseModel `bun:"table:projects"`
+
+	// immutable attributes
+	ID int `bun:"id,pk" json:"id"`
+
+	// mutable attributes
+	Name        string      `bun:"name" json:"name"`
+	Description string      `bun:"description" json:"description"`
+	Archived    bool        `bun:"archived" json:"archived"`
+	CreatedAt   time.Time   `bun:"created_at" json:"created_at"`
+	Notes       JSONB       `bun:"notes" json:"notes"`
+	WorkspaceID int         `bun:"workspace_id" json:"workspace_id"`
+	UserID      int         `bun:"user_id" json:"user_id"`
+	Immutable   bool        `bun:"immutable" json:"immutable"`
+	State       model.State `bun:"state" json:"state"`
+
+	// metadata
+	Seq int64 `bun:"seq" json:"seq"`
+}
+
+// SeqNum gets the SeqNum from a ProjectMsg.
+func (pm *ProjectMsg) SeqNum() int64 {
+	return pm.Seq
+}
+
+// UpsertMsg creates a Project stream upsert message.
+func (pm *ProjectMsg) UpsertMsg() stream.UpsertMsg {
+	return stream.UpsertMsg{
+		JSONKey: ProjectsUpsertKey,
+		Msg:     pm,
+	}
+}
+
+// DeleteMsg creates a Project stream delete message.
+func (pm *ProjectMsg) DeleteMsg() stream.DeleteMsg {
+	deleted := strconv.FormatInt(int64(pm.ID), 10)
+	return stream.DeleteMsg{
+		Key:     ProjectsDeleteKey,
+		Deleted: deleted,
+	}
+}
+
+// ProjectSubscriptionSpec is what a user submits to define a project subscription.
+type ProjectSubscriptionSpec struct {
+	WorkspaceIDs []int `json:"workspace_ids"`
+	ProjectIDs   []int `json:"project_ids"`
+	Since        int64 `json:"since"`
+}
+
+// ProjectCollectStartupMsgs collects ProjectMsg's that were missed prior to startup.
+func ProjectCollectStartupMsgs(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec ProjectSubscriptionSpec,
+) (
+	[]stream.MarshallableMsg, error,
+) {
+	var out []stream.MarshallableMsg
+
+	if len(spec.WorkspaceIDs) == 0 && len(spec.ProjectIDs) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     ProjectsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetProjectStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	var data []*ProjectMsg
+	q := db.Bun().NewSelect().Model(&data).Order("id ASC")
+	q = permFilterQuery(q, accessMap, accessScopes)
+
+	// Ignore tmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.WorkspaceIDs) > 0 {
+		ws.Include("workspace_id in (?)", bun.In(spec.WorkspaceIDs))
+	}
+	if len(spec.ProjectIDs) > 0 {
+		ws.Include("id in (?)", bun.In(spec.ProjectIDs))
+	}
+	q = ws.Apply(q)
```

Comment:
question: this is different than my prototype... in my prototype I selected just the IDs here and then in a later step selected the full trials from the database.  The motivation for that was to reduce the total traffic that the database needed to send to only what was completely necessary.  What is the motivation for this strategy?

Reply:
It looks like the motivation for reducing database got crossed with sourcing the workspace_id in the implementation of project streaming, I'll get this corrected back to the original approach.